File: SplitText/SplitText.java
import java.util.*;

public class SplitText {

    public static void main(String[] args){
	   Set<String> dict = new HashSet<String>();
	   dict.add("he");
	   dict.add("hello");
	   dict.add("hell");
	   dict.add("other");
	   dict.add("there");
	   String word = "hellother";
	   String wordList = create(word, dict, word.length());
	   System.out.println(wordList);

    }

    public static String create(String leftWord, Set<String> dictionary, int oLength){
    	if (dictionary.contains(leftWord))
    	    return leftWord + " ";

	for (int i = 1; i < leftWord.length(); i++){
	    String answer = create(leftWord.substring(0, i), dictionary, oLength) + create(leftWord.substring(i, leftWord.length()), dictionary, oLength);
	    if (valueSum(answer) == oLength)
	    	return answer;
	}
	
	return new String();

    }

    public static int valueSum(String word){
    	int sum = 0;
    	for (int i = 0; i < word.length(); i++){
	    if (word.charAt(i) != ' ')
	    	sum++;
	}

	return sum;
    }

}

File: SplitText/SplitText.py
# SplitText.py

def main():
    dictionary = set()
    dictionary.add('hell')
    dictionary.add('hello')
    dictionary.add('other')
    dictionary.add('there')
    word = 'hellother'
    answer = split_text(word, dictionary, len(word))
    print answer

def split_text(word, dictionary, length):
    if dictionary.__contains__(word):
        return word + str(" ")

    for i in range(1, len(word)):
        candidate = split_text(word[:i], dictionary, length) + \
                split_text(word[i:], dictionary, length)

        if value_sum(candidate) == length:
            return candidate

    return ""

def value_sum(word):
    value = 0
    for i in range(len(word)):
        if word[i] != " ":
            value += 1

    return value

main()
